src/main/java/com/es/phoneshop/service/implementation/ProductServiceImpl.java

```diff
@@ -0,0 +1,78 @@
+package com.es.phoneshop.service.implementation;
+
+import com.es.phoneshop.dao.ProductDao;
+import com.es.phoneshop.dao.implementation.ArrayListProductDaoImpl;
+import com.es.phoneshop.model.Product;
+import com.es.phoneshop.service.ProductService;
+import lombok.NonNull;
+
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+public class ProductServiceImpl implements ProductService {
+    private final ProductDao productDao;
+    private final ReadWriteLock lock;
+
+    public ProductServiceImpl() {
+        productDao = new ArrayListProductDaoImpl();
+        lock = new ReentrantReadWriteLock();
```

Comment:
We should provide thread safety at the lowest layer, where we have the direct access to the list. In our case it should be implemented in dao class to make its methods thread safe and then use it without any risk in any services